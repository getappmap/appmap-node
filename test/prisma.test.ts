import { integrationTest, readAppmap, resolveTarget, runAppmapNode } from "./helpers";
import { copyFileSync, rmSync } from "node:fs";
import { spawnSync } from "node:child_process";

integrationTest("mapping Prisma tests", () => {
  const testDir = resolveTarget();
  copyFileSync(`${testDir}/original.db`, `${testDir}/test.db`);
  spawnSync("yarn", ["prisma", "generate"], { cwd: testDir, shell: true });
  expect(runAppmapNode("yarn", "node", "script.js").status).toBe(0);
  expect(readAppmap()).toMatchSnapshot();
  rmSync(`${testDir}/test.db`);
});
