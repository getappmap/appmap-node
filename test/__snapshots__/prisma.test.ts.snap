// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapping Prisma tests 1`] = `
{
  "classMap": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "location": "@prisma/client/Post:2",
                      "name": "deleteMany",
                      "static": true,
                      "type": "function",
                    },
                  ],
                  "name": "Post",
                  "type": "class",
                },
                {
                  "children": [
                    {
                      "location": "@prisma/client/User:1",
                      "name": "findMany",
                      "static": true,
                      "type": "function",
                    },
                    {
                      "location": "@prisma/client/User:3",
                      "name": "deleteMany",
                      "static": true,
                      "type": "function",
                    },
                    {
                      "location": "@prisma/client/User:4",
                      "name": "create",
                      "static": true,
                      "type": "function",
                    },
                  ],
                  "name": "User",
                  "type": "class",
                },
              ],
              "name": "client",
              "type": "package",
            },
          ],
          "name": "@prisma",
          "type": "package",
        },
        {
          "children": [
            {
              "location": "script.js:6",
              "name": "main",
              "static": true,
              "type": "function",
            },
          ],
          "name": "script",
          "type": "class",
        },
      ],
      "name": "prisma-appmap-node-test",
      "type": "package",
    },
  ],
  "events": [
    {
      "defined_class": "script",
      "event": "call",
      "id": 1,
      "lineno": 6,
      "method_id": "main",
      "parameters": [],
      "path": "script.js",
      "static": true,
      "thread_id": 0,
    },
    {
      "defined_class": "User",
      "event": "call",
      "id": 3,
      "lineno": 1,
      "method_id": "findMany",
      "parameters": [
        {
          "class": "Object",
          "name": "args",
          "object_id": 2,
          "properties": [
            {
              "class": "Object",
              "name": "where",
              "properties": [
                {
                  "class": "Object",
                  "name": "name",
                  "properties": [
                    {
                      "class": "String",
                      "name": "contains",
                    },
                  ],
                },
              ],
            },
          ],
          "value": "{ where: { name: { contains: 'Admin' } } }",
        },
      ],
      "path": "@prisma/client/User",
      "receiver": {
        "class": "String",
        "value": "'User'",
      },
      "static": false,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 4,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "SELECT \`main\`.\`User\`.\`id\`, \`main\`.\`User\`.\`email\`, \`main\`.\`User\`.\`name\` FROM \`main\`.\`User\` WHERE (\`main\`.\`User\`.\`name\` LIKE ? AND \`main\`.\`User\`.\`email\` LIKE ?) LIMIT ? OFFSET ?",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 5,
      "parent_id": 4,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 6,
      "parent_id": 3,
      "return_value": {
        "class": "Promise<Array>",
        "object_id": 3,
        "value": "Promise { [] }",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "Post",
      "event": "call",
      "id": 7,
      "lineno": 2,
      "method_id": "deleteMany",
      "parameters": [
        {
          "class": "undefined",
          "name": "args",
          "value": "undefined",
        },
      ],
      "path": "@prisma/client/Post",
      "receiver": {
        "class": "String",
        "value": "'Post'",
      },
      "static": false,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 8,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "DELETE FROM \`main\`.\`Post\` WHERE 1=1",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 9,
      "parent_id": 8,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 10,
      "parent_id": 7,
      "return_value": {
        "class": "Promise<Object>",
        "object_id": 4,
        "value": "Promise { { count: 0 } }",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "User",
      "event": "call",
      "id": 11,
      "lineno": 3,
      "method_id": "deleteMany",
      "parameters": [
        {
          "class": "undefined",
          "name": "args",
          "value": "undefined",
        },
      ],
      "path": "@prisma/client/User",
      "receiver": {
        "class": "String",
        "value": "'User'",
      },
      "static": false,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 12,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "DELETE FROM \`main\`.\`User\` WHERE 1=1",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 13,
      "parent_id": 12,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 14,
      "parent_id": 11,
      "return_value": {
        "class": "Promise<Object>",
        "object_id": 5,
        "value": "Promise { { count: 0 } }",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "User",
      "event": "call",
      "id": 19,
      "lineno": 4,
      "method_id": "create",
      "parameters": [
        {
          "class": "Object",
          "name": "args",
          "object_id": 6,
          "properties": [
            {
              "class": "Object",
              "name": "data",
              "properties": [
                {
                  "class": "String",
                  "name": "name",
                },
                {
                  "class": "String",
                  "name": "email",
                },
              ],
            },
          ],
          "value": "{ data: { name: 'Alice', email: 'alice@prisma.io' } }",
        },
      ],
      "path": "@prisma/client/User",
      "receiver": {
        "class": "String",
        "value": "'User'",
      },
      "static": false,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 20,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "INSERT INTO \`main\`.\`User\` (\`email\`, \`name\`) VALUES (?,?) RETURNING \`id\` AS \`id\`, \`email\` AS \`email\`, \`name\` AS \`name\`",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 21,
      "parent_id": 20,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 22,
      "parent_id": 19,
      "return_value": {
        "class": "Promise<Object>",
        "object_id": 7,
        "value": "Promise { { id: 1, email: 'alice@prisma.io', name: 'Alice' } }",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "User",
      "event": "call",
      "id": 23,
      "lineno": 4,
      "method_id": "create",
      "parameters": [
        {
          "class": "Object",
          "name": "args",
          "object_id": 8,
          "properties": [
            {
              "class": "Object",
              "name": "data",
              "properties": [
                {
                  "class": "String",
                  "name": "name",
                },
                {
                  "class": "String",
                  "name": "email",
                },
                {
                  "class": "Object",
                  "name": "posts",
                  "properties": [
                    {
                      "class": "Object",
                      "name": "create",
                      "properties": [
                        {
                          "class": "String",
                          "name": "title",
                        },
                      ],
                    },
                  ],
                },
              ],
            },
          ],
          "value": "{
  data: {
    name: 'Bob',
    email: 'bob@prisma.io',
    posts: { create: { title: 'Hello World' } }
  }
}",
        },
      ],
      "path": "@prisma/client/User",
      "receiver": {
        "class": "String",
        "value": "'User'",
      },
      "static": false,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 24,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "BEGIN",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 25,
      "parent_id": 24,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 26,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "INSERT INTO \`main\`.\`User\` (\`email\`, \`name\`) VALUES (?,?) RETURNING \`id\` AS \`id\`",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 27,
      "parent_id": 26,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 28,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "INSERT INTO \`main\`.\`Post\` (\`title\`, \`published\`, \`authorId\`) VALUES (?,?,?) RETURNING \`id\` AS \`id\`",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 29,
      "parent_id": 28,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 30,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "SELECT \`main\`.\`User\`.\`id\`, \`main\`.\`User\`.\`email\`, \`main\`.\`User\`.\`name\` FROM \`main\`.\`User\` WHERE \`main\`.\`User\`.\`id\` = ? LIMIT ? OFFSET ?",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 31,
      "parent_id": 30,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 32,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "COMMIT",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 33,
      "parent_id": 32,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 34,
      "parent_id": 23,
      "return_value": {
        "class": "Promise<Object>",
        "object_id": 9,
        "value": "Promise { { id: 2, email: 'bob@prisma.io', name: 'Bob' } }",
      },
      "thread_id": 0,
    },
    {
      "defined_class": "User",
      "event": "call",
      "id": 35,
      "lineno": 1,
      "method_id": "findMany",
      "parameters": [
        {
          "class": "Object",
          "name": "args",
          "object_id": 10,
          "properties": [
            {
              "class": "Object",
              "name": "include",
              "properties": [
                {
                  "class": "Boolean",
                  "name": "posts",
                },
              ],
            },
            {
              "class": "Object",
              "name": "where",
              "properties": [
                {
                  "class": "Object",
                  "name": "name",
                  "properties": [
                    {
                      "class": "String",
                      "name": "contains",
                    },
                  ],
                },
              ],
            },
          ],
          "value": "{ include: { posts: true }, where: { name: { contains: 'Bob' } } }",
        },
      ],
      "path": "@prisma/client/User",
      "receiver": {
        "class": "String",
        "value": "'User'",
      },
      "static": false,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 36,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "SELECT \`main\`.\`User\`.\`id\`, \`main\`.\`User\`.\`email\`, \`main\`.\`User\`.\`name\` FROM \`main\`.\`User\` WHERE \`main\`.\`User\`.\`name\` LIKE ? LIMIT ? OFFSET ?",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 37,
      "parent_id": 36,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 38,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "SELECT \`main\`.\`Post\`.\`id\`, \`main\`.\`Post\`.\`title\`, \`main\`.\`Post\`.\`content\`, \`main\`.\`Post\`.\`published\`, \`main\`.\`Post\`.\`authorId\` FROM \`main\`.\`Post\` WHERE \`main\`.\`Post\`.\`authorId\` IN (?) LIMIT ? OFFSET ?",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 39,
      "parent_id": 38,
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 40,
      "parent_id": 35,
      "return_value": {
        "class": "Promise<Array>",
        "object_id": 11,
        "value": "Promise { [ { id: 2, email: 'bob@prisma.io', name: 'Bob', posts: [Array] } ] }",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 2,
      "parent_id": 1,
      "return_value": {
        "class": "Promise<undefined>",
        "object_id": 1,
        "value": "Promise { undefined }",
      },
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 15,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "delete from sqlite_sequence where name='Post'",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 16,
      "parent_id": 15,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 17,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "delete from sqlite_sequence where name='User'",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 18,
      "parent_id": 17,
      "thread_id": 0,
    },
    {
      "event": "call",
      "id": 41,
      "sql_query": {
        "database_type": "sqlite",
        "sql": "SELECT 1",
      },
      "thread_id": 0,
    },
    {
      "elapsed": 31.337,
      "event": "return",
      "id": 42,
      "parent_id": 41,
      "thread_id": 0,
    },
  ],
  "metadata": {
    "app": "prisma-appmap-node-test",
    "client": {
      "name": "appmap-node",
      "url": "https://github.com/getappmap/appmap-node",
      "version": "test node-appmap version",
    },
    "language": {
      "engine": "Node.js",
      "name": "javascript",
      "version": "test node version",
    },
    "name": "test process recording",
    "recorder": {
      "name": "process",
      "type": "process",
    },
  },
  "version": "1.12",
}
`;
